casper/src/main/scala/coop/rchain/casper/TuranOracle.scala

```diff
@@ -0,0 +1,103 @@
+import util._
+
+import scala.collection
+
+/*
+ * Implementation inspired by Ethereum's CBC casper simulator's Turan oracle implementation.
+ *
+ * https://github.com/ethereum/cbc-casper/blob/0.2.0/casper/safety_oracles/turan_oracle.py
+ *
+ * "If nodes in an e-clique see each other agreeing on e and can't see each other disagreeing on e,
+ * then there does not exist any new message from inside the clique that will cause them to assign
+ * lower scores to e. Further, if the clique has more than half of the validators by weight,
+ * then no messages external to the clique can raise the scores these validators assign to
+ * a competing estimate to be higher than the score they assign to e."
+ *
+ * - From https://github.com/ethereum/research/blob/master/papers/CasperTFG/CasperTFG.pdf
+ *
+ * That is unless there are equivocations.
+ * The fault tolerance threshold is a subjective value that the user sets to "secretly" state that they
+ * tolerate up to fault_tolerance_threshold fraction of the total weight to equivocate.
+ *
+ * In the extreme case when your fault tolerance threshold is 1,
+ * all validators must be part of the clique that supports the estimate in order to state that it is finalized.
+ * If all validators are indeed part of the clique, min_max_clique_weight will hopefully be equal to total_weight
+ * and is_safe will reduce to total_weight >= total_weight and evaluate to true.
+ */
+class TuranOracle(blocks: collection.Map[ByteString, BlockMessage],
+                  estimate: BlockMessage,
+                  latestBlocks: collection.Map[ByteString, BlockMessage],
+                  fault_tolerance_threshold: Float) {
+  def totalWeight(): Int = weightMapTotal(estimate)
+  def isSafe: Boolean = {
+    val fault_tolerance = 2 * min_max_clique_weight() - totalWeight()
+    fault_tolerance >= fault_tolerance_threshold * totalWeight()
+  }
+
+  private def min_max_clique_weight(): Int =
+    // To have a maximum clique of half the total weight, you need at least twice the weight of the candidates to be greater than the total weight
+    if (2 * candidates().values.sum < totalWeight()) {
+      0
+    } else {
+      val vertexCount = candidates().keys.size
+      val edgeCount   = agreementGraphEdgeCount(blocks, estimate, latestBlocks, candidates())
+      minTotalValidatorWeight(estimate, maxCliqueMinSize(vertexCount, edgeCount))
+    }
+
+  private def candidates(): Map[ByteString, Int] = {
+    val estimateMainParent = mainParent(blocks, estimate)
+    val weights = estimateMainParent match {
+      case Some(parent) => weightMap(parent)
+      case None         => weightMap(estimate) // Genesis
+    }
+    for {
+      (validator, stake) <- weights
+      latestBlock        <- latestBlocks.get(validator) if compatible(estimate, latestBlock)
+    } yield (validator, stake)
+  }
+
+  // TODO: Add free messages
+  def agreementGraphEdgeCount(blocks: collection.Map[ByteString, BlockMessage],
```

Comment:
This function takes many of the same arguments as the class constructor. Would it make sense to remove them from the method call here and make them `val`s of the class instead? E.g. `class TuranOracle(val blocks: ...)`.